### PR TITLE
Kyma modularization proposal

### PR DESCRIPTION
**Pull request by PK85**
**Description**

Changes proposed in this pull request:

- The Kyma modularization proposal to allow users to add/remove/replace kyma components in more flexible way during the installation and in runtime

**Related issue(s)**

### CODE REVIEW

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+## Goal
+
+Allow users to add/remove/replace Kyma components in more flexible way during the installation and in runtime. This should be possible:
+
+- install Kyma lite on minikube (without monitoring, console and ui-api-layer, tracing, service catalog, and remote environments)
+- install Kyma without service catalog and brokers installed on GKE with already provided Service Catalog and GCP broker included
+- install Kyma as gardener addon Gardener with dependencies to Istio and Knative addons
+- add service broker to existing Kyma installation in runtime (e.g. azure broker)
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `a Service Broker`
~(e.g. azure broker)~ `, for example Azure Broker`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+- add new remote environment (also new type of application) in runtime
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `a new Remote Environment`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+- add completely new Kyma component that integrates with UI in runtime (register microfrontend and custom resources)
+- use user provided S3 bucket instead of minio
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `Minio`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+- install private modules (from private docker registry or private git repository)
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `Docker`
`Git`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+- install components from Kyma examples/showcases catalog easily (UI)
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `examples and showcases`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+- new installation parameters or dependencies require code changes
+
+## Proposal
+
+Flexible and modular installation of fine grained components with limited dependencies. The components installation and update should be more generic and declarative. This is the first list of ideas how to improve current solution:
+
+- parameters should not be hard coded - there should be a convention how to pass parameters through installation process to Kyma components
+- default values everywhere - the list of mandatory parameters should be as short as possible (the goal: only cluster domain is a mandatory parameter)
+- optional dependencies, e.g. Kyma ui-api-layer can work without custom resources introduced by service catalog or remote environment - still can support other areas.
+- dynamic UI (microfrontends) + back-end - adding new components extend views (but also extend backed functionality like supported graphQL queries)
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: (micro front-ends) ?
`adding new components extends views and also the backed functionalities, such as the supported GraphQL queries.`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+- no custom logic in installation process (shell scripts) - it required it should be replaced by the operator pattern
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `the installation process`
~it required~ `it should be replaced by the operator pattern.`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+- set minimal set of mandatory components that form real Kyma "core" (the dependency for core can be mandatory, other dependencies should be optional)
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `The minimal set of mandatory components that form the real Kyma "core". The dependency for the core can be mandatory, other dependencies should be optional.`

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+- Kyma core as gardener addon (see: https://github.com/knative/docs/blob/master/install/Knative-with-Gardener.md#alternative-way-to-enable-knative-with-gardener and https://github.com/gardener/bouquet)
```
Comment (klaudiagrz) [inline sig-and-wg/sig-core/proposals/modularization.md]: `as a Gardener addon`

I would look better if the links were hidden. For example:
``For reference, see [Knative](https://github.com/knative/docs/blob/master/install/Knative-with-Gardener.md#alternative-way-to-enable-knative-with-gardener) and [Gardener](https://github.com/gardener/bouquet) documentation.``

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+- split Kyma core chart into more fine grained components, and provide a tooling to install all of them in the same convenient way as before (convention)
+- Kyma addons/brokers/showcases as a Kyma bundles (see POC: https://github.com/PK85/kyma-bundles )
+- Advanced Level: acceptance testing, especially ui-api-layer and cross components tests
+
+### Modularity - refactoring
+
+![](assets/modularization.svg)
+
+### Kyma bundles for service catalog
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+This project:
+
+- allows you to add your expanded helm broker bundle, then
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 9 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+- install private modules (from private docker registry or private git repository)
+- install components from Kyma examples/showcases catalog easily (UI)
+
+## Current solution
+
+Current installation process is a mix of:
+
+- custom shell scripts, 
+- kubectl commands,
+- helm charts including one big, fat core chart,
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+## Current solution
+
+Current installation process is a mix of:
+
+- custom shell scripts, 
+- kubectl commands,
+- helm charts including one big, fat core chart, 
+- complex parametrization
+
+Installer is a kubernetes operator with hard coded logic executing all the components above. Problems with this solution:
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+- default minikube installation is slow and require a lot of resources
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+- shell scripts are difficult to manage (some of them have to support mac, linux, windows)
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+Created on 2018-08-16 by Piotr Kopczynski (@PK85) and Piotr Bochyński (@pbochynski).
+
+## Status
+
+Proposed on 2018-08-16.
+
+## Motivation
+
+Kyma provides components as a one big pack to enable developers to build something on cloud native stack. They work each other well and together expose unified developer and business experience. Now, we need to go one step back and enable users to decide what they want to use from Kyma.
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+The landscape around Kyma has changed. Kubernetes managed clusters get popular. Now each big cloud infrastructure provider has own managed kubernetes service (Google GKE, Amazon EKS, Microsoft AKS). This offering is evolving and customers can use more services included in the managed kubernetes (like Istio - service mesh, service catalog, logging). Developers work on bringing more features to transform kubernetes from container scheduling system into complete platform solution. Kyma is not the only solution in this area. Knative is the new open source project started by Google that also aims to provide components to deliver cloud native software on top of kubernetes. There is strategic decision to run Kyma on top of knative. All this development in the kubernetes landscape shows the need of having flexible loosely coupled components to enable customers to use best available solutions from infrastructure providers, open source projects and commercial solutions. Kyma in the principle offers complete solution (batteries included), but it doesn't mean Kyma can't benefit of running on top of managed cluster, by integration with already provided services.
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+## Goal
+
+Allow users to add/remove/replace Kyma components in more flexible way during the installation and in runtime. This should be possible:
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+- install Kyma lite on minikube (without monitoring, console and ui-api-layer, tracing, service catalog, and remote environments)
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 1 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 2 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+- default minikube installation is slow and require a lot of resources
+- shell scripts are difficult to manage (some of them have to support mac, linux, windows)
+- new installation parameters or dependencies require code changes
+
+## Proposal
+
+Flexible and modular installation of fine grained components with limited dependencies. The components installation and update should be more generic and declarative. This is the first list of ideas how to improve current solution:
+
+- parameters should not be hard coded - there should be a convention how to pass parameters through installation process to Kyma components
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+- default values everywhere - the list of mandatory parameters should be as short as possible (the goal: only cluster domain is a mandatory parameter)
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+- optional dependencies, e.g. Kyma ui-api-layer can work without custom resources introduced by service catalog or remote environment - still can support other areas.
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+- dynamic UI (microfrontends) + back-end - adding new components extend views (but also extend backed functionality like supported graphQL queries)
+- no custom logic in installation process (shell scripts) - it required it should be replaced by the operator pattern
+- set minimal set of mandatory components that form real Kyma "core" (the dependency for core can be mandatory, other dependencies should be optional)
+- Kyma core as gardener addon (see: https://github.com/knative/docs/blob/master/install/Knative-with-Gardener.md#alternative-way-to-enable-knative-with-gardener and https://github.com/gardener/bouquet)
+- split Kyma core chart into more fine grained components, and provide a tooling to install all of them in the same convenient way as before (convention)
+- Kyma addons/brokers/showcases as a Kyma bundles (see POC: https://github.com/PK85/kyma-bundles )
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+
+## Goal
+
+Allow users to add/remove/replace Kyma components in more flexible way during the installation and in runtime. This should be possible:
+
+- install Kyma lite on minikube (without monitoring, console and ui-api-layer, tracing, service catalog, and remote environments)
+- install Kyma without service catalog and brokers installed on GKE with already provided Service Catalog and GCP broker included
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 3 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 4 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+
+## Goal
+
+Allow users to add, remove, and replace Kyma components in more flexible way during the installation and in runtime. The following processes should be possible:
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+- Install Kyma lite on Minikube (without Monitoring, console and ui-api-layer, Tracing, Service Catalog, and Remote Environments).
+- Install Kyma without the Service Catalog and brokers installed on GKE with already provided Service Catalog and GCP broker included.
+- Install Kyma as gardener addon Gardener with dependencies to Istio and Knative addons.
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+- Add completely new Kyma component that integrates with UI in runtime (register micro front-end and custom resources).
+- Use user provided S3 bucket instead of Minio.
+- Install private modules (from private Docker registry or private Git repository).
+- Install components from Kyma examples and showcases catalog easily (UI).
+
+## Current solution
+
+Current installation process is a mix of:
+
+- Custom shell scripts.
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+Installer is a Kubernetes operator with hard-coded logic executing all the components above. Problems with this solution are as follows:
+
+- The default Minikube installation is slow and requires a lot of resources.
+- Shell scripts are difficult to manage (some of them have to support Mac, Linux, Windows).
+- New installation parameters or dependencies require code changes.
+
+## Proposal
+
+Flexible and modular installation of fine grained components with limited dependencies. The components installation and update should be more generic and declarative. This is the first set of ideas on how to improve the current solution:
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+### Modularity - refactoring
+
+![](assets/modularization.svg)
+
+### Kyma bundles for the Service Catalog
+
+This [project](https://github.com/PK85/kyma-bundles):
+
+- Allows you to add your expanded Helm Broker bundle.
+- Will take care of updating bundle repository(http server) with generated .tgz files and updated index.yaml.
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+- Optional dependencies, for example: Kyma ui-api-layer can work without custom resources introduced by the Service Catalog or Remote Environment and still can support other areas.
+- Dynamic UI (micro front-ends) + back-end - adding new components extends views and also the backed functionalities, such as the supported GraphQL queries.
+- No custom logic in the installation process (shell scripts) - it should be replaced by the operator pattern.
+- The minimal set of mandatory components that form the real Kyma "core". The dependency for the core can be mandatory, other dependencies should be optional.
+- Kyma core as Gardener addon. For reference, see [Knative with Gardener](https://github.com/knative/docs/blob/master/install/Knative-with-Gardener.md#alternative-way-to-enable-knative-with-gardener) and [Gardener Bouquet](https://github.com/gardener/bouquet) documentation.
+- Split Kyma core chart into more fine grained components, and provide a tooling to install all of them in the same convenient way as before (convention).
+- Kyma addons/brokers/showcases as Kyma bundles. For reference, see: [Kyma bundles](https://github.com/PK85/kyma-bundles).
```
[the remark this answers is not in the document]
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+
+The landscape around Kyma has changed. Kubernetes managed clusters get popular. Each big cloud infrastructure provider has own managed Kubernetes service (Google GKE, Amazon EKS, Microsoft AKS). This offering is evolving and customers can use more services included in the managed Kubernetes (like Istio - Service Mesh, Service Catalog, Logging). Developers work on bringing more features to transform Kubernetes from container scheduling system into complete platform solution. Kyma is not the only solution in this area. Knative is the new open source project started by Google that also aims to provide components to deliver cloud native software on top of Kubernetes. There is a strategic decision to run Kyma on top of Knative. All this development in the Kubernetes landscape shows the need of having flexible, loosely-coupled components to enable customers using best available solutions from infrastructure providers, open source projects and commercial solutions. In principle, Kyma offers a complete solution (batteries included), but it doesn't mean Kyma can't benefit from running on top of managed cluster, by integration with the already provided services.
+
+## Goal
```
Comment (derberg) [inline sig-and-wg/sig-core/proposals/modularization.md]: Additional goal is to: Improve speed of testing components and integrating them in Kyma. We need to be able to install kyma with only components that are modified and run only acceptance tests related to them.

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+- Dynamic UI (micro front-ends) + back-end - adding new components extends views and also the backed functionalities, such as the supported GraphQL queries.
+- No custom logic in the installation process (shell scripts) - it should be replaced by the operator pattern.
+- The minimal set of mandatory components that form the real Kyma "core". The dependency for the core can be mandatory, other dependencies should be optional.
+- Kyma core as Gardener add-on. For reference, see [Knative with Gardener](https://github.com/knative/docs/blob/master/install/Knative-with-Gardener.md#alternative-way-to-enable-knative-with-gardener) and [Gardener Bouquet](https://github.com/gardener/bouquet) documentation.
+- Split Kyma core chart into more fine grained components, and provide a tooling to install all of them in the same convenient way as before (convention).
+- Kyma add-ons, brokers, and showcases as Kyma bundles. For reference, see the [Kyma bundles](https://github.com/PK85/kyma-bundles) repository.
+- Advanced Level: acceptance testing, especially ui-api-layer and cross components tests.
```
Comment (derberg) [inline sig-and-wg/sig-core/proposals/modularization.md]: I would rephrase this sentence:
Split acceptance tests and keep them in the same chart with specific component (Addons). Make sure it is possible to configure what tests is run when, depending on dependency tree.

File: sig-and-wg/sig-core/proposals/modularization.md [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+
+### Modularity - refactoring
+
+![](assets/modularization.svg)
+
+### Kyma bundles for the Service Catalog
```
Comment (derberg) [inline sig-and-wg/sig-core/proposals/modularization.md]: This is one of sub-proposals and should not in my opinion be present in this proposal

Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: I'm going to extract that in a next phase, already decided that we will go with kyma bundles for brokers and showcases.

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 32 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: That sentence was added by you, and you added review comment to your sentence. 
The dependency tree and mandatory/optional components proposal needs to be provided.

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 31 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: It just matters for local installation, right?

Comment (derberg) [reply sig-and-wg/sig-core/proposals/modularization.md]: for now we can say it matters for local as it would be mainly used for development on local, or for PRs within docker-in-docker. But if the modularization is effective and GKE and other providers fast enough, maybe we can stop maintaining docker-in-docker and do proper testing on a cluster for a PR - just an idea how we can profit in future

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 32 above)
Comment (derberg) [reply sig-and-wg/sig-core/proposals/modularization.md]: 🙇

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 33 above)
Comment (derberg) [reply sig-and-wg/sig-core/proposals/modularization.md]: maybe decided we do it, but proposal will explain how
anyway, not part of this PR ;)

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 32 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done

File: sig-and-wg/sig-core/proposals/modularization.md (hunk 33 above)
Comment (PK85) [reply sig-and-wg/sig-core/proposals/modularization.md]: done
